feat: allow copying service name from service/dataplane list views

kleinfreund opened the pull request:
**feat(services): add copy button to service name in list view**

**refactor(CopyButton): use type syntax for props**

**feat(dataplanes): show copy button for services in list view**

Add support for showing a copy button to the TagList component.

Resolves #1747.

Signed-off-by: Philipp Rudloff <philipp.rudloff@konghq.com>

johncowen reviewed (commented):
I just took a very quick look at this, and clicking on the copy button in the listing view navigates me to the detail view.

Apart from that I'm not totally sold on the copy button being in the tag thing, but as I said I just took a very quick look and noticed the above thing, so leave that one with me some more.

kleinfreund commented:
> clicking on the copy button in the listing view navigates me to the detail view.

Good catch. I’ve fixed this now. We had this before and the strategy was making sure the links are nested inside `TextWithCopyButton` and not the other way around.

> I'm not totally sold on the copy button being in the tag thing, but as I said I just took a very quick look and noticed the above thing, so leave that one with me some more.

Me neither. I was thinking to stop using the TagList component specifically for the Dataplane list view. Let me try that. That might just be the solution here.

kleinfreund commented:
> I was thinking to stop using the TagList component specifically for the Dataplane list view. Let me try that. That might just be the solution here.

I’ve implemented this now. I think I like this better.

kleinfreund commented:
> Oh, one thing I noticed which might not be fixable easily here (but could be, lmk) is the non-rounded corners on the tooltip/notification things:

I added the Kongponents default border radius to this now. I have work planned to replace CopyButton with KCopy completely, but that’s not priority right now.

kleinfreund commented:
> I added the Kongponents default border radius to this now. I have work planned to replace CopyButton with KCopy completely, but that’s not priority right now.

Huh, I didn’t realize this already had a border radius. Let me look at this some more.

Okay, so specifically it’s not applied for the services in the Dataplane list. Strange. I’ll look at this a bit. If it’s not easy to fix, let’s do that separately.

kleinfreund commented:
Okay, yes, let me look at that separately. The issue is the KTruncate use using `overflow: hidden` which cuts-off parts of the tooltip making it look like it has no border-radius. That’s not exactly trivial to fix.

johncowen commented:
> That’s not exactly trivial to fix.

Cool don't worry on that one then thanks for looking.

I noticed this also, looks like the service names have turned blue in this PR:

![Screenshot 2023-12-07 at 15 13 53](https://github.com/kumahq/kuma-gui/assets/554604/d920e085-7f1b-4ab7-9815-6b3668170924)

kleinfreund commented:
> I noticed this also, looks like the service names have turned blue in this PR:

I changed the AppCollection styles for this to fix that and also removed an unnecessary `text-decoration: none;`.

johncowen commented:
> and also removed an unnecessary text-decoration: none;.

Oh wait, that means we get an underline on hover which I don't think we want

kleinfreund commented:
> Oh wait, that means we get an underline on hover which I don't think we want

Isn’t that _exactly_ what we want? We have the following base rule for links:

```css
a:hover,
a:focus {
  text-decoration: underline
}
```

So why shouldn’t it apply to the name cells in AppCollection?

johncowen commented:
It's not what we want (unless something has changed that I'm not aware of). We spoke about this a good while ago now.

The name of the thing in the row (i.e. the first cell) shouldn't look/act like a link, the entire row is the clickable thing. The text in the first cell shouldn't be blue/underlined/underline on hover or anything that makes it look like a link. It should be black and slightly bold, as close as possible to this:

<img width="126" alt="Screenshot 2023-12-08 at 10 18 07" src="https://github.com/kumahq/kuma-gui/assets/554604/394dea86-461e-4115-8e4a-b11ed6a6b3b2">

> We have the following base rule for links:

This is why this rule is there, to overwrite the link base rule.

kleinfreund commented:
> It's not what we want (unless something has changed that I'm not aware of). We spoke about this a good while ago now.
> 
> The name of the thing in the row (i.e. the first cell) shouldn't look/act like a link, the entire row is the clickable thing. The text in the first cell shouldn't be blue/underlined/underline on hover or anything that makes it look like a link. It should be black and slightly bold, as close as possible to this:

I’ve set this declaration again as it’s not related to this PR’s purpose, but I really, really disagree with this entire statement.

It’s not clear why something that’s a link shouldn’t look like any other link in the application by default. In this case, a special rule applies to the color and font-weight as a first cell item in tables, but why the link shouldn’t be underlined on hover is unclear.

The link also very much acts like a link: You can activate it to navigate, you can focus it, you can use the whole array of browser link interactions like MouseWheel+Click, Ctrl+Click, etc.

johncowen commented on src/app/application/components/app-collection/AppCollection.vue:
```diff
@@ -225,10 +225,13 @@ const click = (e: MouseEvent) => {
+}
+
+.app-collection :deep(td:first-child a) {
+  color: currentColor;
```
I don't think there are any changes required in this file. You can just leave it as it is. I think this has only ended up being changed because of the little visual changes made here which we've now undone, but we've undone them differently to how they were in the first place instead of just undoing.

Another note: `color: currentColor` is a little weird seeing as we are setting the `color` to the current color, which we are setting to the current color, which we are setting to the current color etc etc. I think `inherit` is what we mean here.

kleinfreund replied:
The changes are required because otherwise the color of the name cell link for services is now blue.

I’m going to leave this as-is because it works.

johncowen replied:
Oh I see, gotcha! Could we just do this then:

```css
.app-collection :deep(td:first-child),
.app-collection :deep(td:first-child *) {
  color: inherit;
  font-weight: $kui-font-weight-semibold;
  text-decoration: none;
}
```

i.e. remove the immediate child selector `>`? Its the removal of this in your above code that solves the issue I think not the change to use `a` or the weird `color: currentColor` (I'd rather not use that if poss, this is my biggest "what the what!?")

kleinfreund replied:
`currentColor` isn’t weird. I encourage you to read up on it and let’s not dwell on PR suggestions that end up being nitpicks. This wasn’t helpful.

I’ve changed this _again_ to what you suggested.

johncowen reviewed (approved):
Thanks!
